{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import flwr as fl\n",
    "from flwr.common.typing import Scalar\n",
    "import ray\n",
    "import torch\n",
    "import torchvision\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from torchvision import transforms\n",
    "from torchvision.datasets import VisionDataset, Food101\n",
    "\n",
    "import numpy as np\n",
    "from collections import OrderedDict\n",
    "from pathlib import Path\n",
    "from typing import Dict, Callable, Optional, Tuple, List, Any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import multiprocessing\n",
    "\n",
    "data_path = os.path.join(os.getcwd(),'data', 'food-101')\n",
    "cpu_count = multiprocessing.cpu_count() - 1 # set as you like!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class LeNet5(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(LeNet5, self).__init__()\n",
    "        \n",
    "        self.convolutional_layer = nn.Sequential(            \n",
    "            nn.Conv2d(in_channels=3, out_channels=20, kernel_size=5, stride=1),\n",
    "            nn.ReLU(),\n",
    "            nn.MaxPool2d(kernel_size=2,stride =2),\n",
    "            nn.Conv2d(in_channels=20, out_channels=50, kernel_size=5, stride=1),\n",
    "            nn.ReLU(),\n",
    "            nn.MaxPool2d(kernel_size=2,stride =2),\n",
    "           \n",
    "        )\n",
    "\n",
    "        self.linear_layer = nn.Sequential(\n",
    "            nn.Linear(in_features=1250, out_features=500),\n",
    "            nn.ReLU(),\n",
    "            nn.Linear(in_features=500, out_features=2),\n",
    "        )\n",
    "\n",
    "\n",
    "    def forward(self, x):\n",
    "        x = self.convolutional_layer(x)\n",
    "        x = torch.flatten(x, 1)\n",
    "        x = self.linear_layer(x)\n",
    "        x = F.softmax(x, dim=1)\n",
    "        return x\n",
    "\n",
    "\n",
    "# borrowed from Pytorch quickstart example\n",
    "def train(net, trainloader, epochs, device: str):\n",
    "    \"\"\"Train the network on the training set.\"\"\"\n",
    "    criterion = torch.nn.CrossEntropyLoss()\n",
    "    optimizer = torch.optim.SGD(net.parameters(), lr=0.01, momentum=0.9)\n",
    "    net.train()\n",
    "    for _ in range(epochs):\n",
    "        for images, labels in trainloader:\n",
    "            images, labels = images.to(device), labels.to(device)\n",
    "            optimizer.zero_grad()\n",
    "            loss = criterion(net(images), labels)\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "\n",
    "\n",
    "# borrowed from Pytorch quickstart example\n",
    "def test(net, testloader, device: str):\n",
    "    \"\"\"Validate the network on the entire test set.\"\"\"\n",
    "    criterion = torch.nn.CrossEntropyLoss()\n",
    "    correct, total, loss = 0, 0, 0.0\n",
    "    net.eval()\n",
    "    with torch.no_grad():\n",
    "        for batch in testloader:\n",
    "            images, labels = batch[0].to(device), batch[1].to(device)\n",
    "            outputs = net(images)\n",
    "            loss += criterion(outputs, labels).item()\n",
    "            _, predicted = torch.max(outputs.data, 1)\n",
    "            total += labels.size(0)\n",
    "            correct += (predicted == labels).sum().item()\n",
    "    accuracy = correct / total\n",
    "    return loss, accuracy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flower client, adapted from Pytorch quickstart example\n",
    "class FlowerClient(fl.client.NumPyClient):\n",
    "    def __init__(self, cid: str, fed_dir_data: str):\n",
    "        self.cid = cid\n",
    "        self.fed_dir = Path(fed_dir_data)\n",
    "        self.properties: Dict[str, Scalar] = {\"tensor_type\": \"numpy.ndarray\"}\n",
    "\n",
    "        # Instantiate model\n",
    "        self.net = LeNet5()\n",
    "\n",
    "        # Determine device\n",
    "        self.device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "    def get_parameters(self, config):\n",
    "        return get_params(self.net)\n",
    "\n",
    "    def fit(self, parameters, config):\n",
    "        set_params(self.net, parameters)\n",
    "\n",
    "        # Load data for this client and get trainloader\n",
    "        num_workers = int(ray.get_runtime_context().get_assigned_resources()[\"CPU\"])\n",
    "        trainloader = get_dataloader(\n",
    "            self.fed_dir,\n",
    "            self.cid,\n",
    "            is_train=True,\n",
    "            batch_size=config[\"batch_size\"],\n",
    "            workers=num_workers,\n",
    "        )\n",
    "\n",
    "        # Send model to device\n",
    "        self.net.to(self.device)\n",
    "\n",
    "        # Train\n",
    "        train(self.net, trainloader, epochs=config[\"epochs\"], device=self.device)\n",
    "\n",
    "        # Return local model and statistics\n",
    "        return get_params(self.net), len(trainloader.dataset), {}\n",
    "\n",
    "    def evaluate(self, parameters, config):\n",
    "        set_params(self.net, parameters)\n",
    "\n",
    "        # Load data for this client and get trainloader\n",
    "        num_workers = int(ray.get_runtime_context().get_assigned_resources()[\"CPU\"])\n",
    "        valloader = get_dataloader(\n",
    "            self.fed_dir, self.cid, is_train=False, batch_size=50, workers=num_workers\n",
    "        )\n",
    "\n",
    "        # Send model to device\n",
    "        self.net.to(self.device)\n",
    "\n",
    "        # Evaluate\n",
    "        loss, accuracy = test(self.net, valloader, device=self.device)\n",
    "\n",
    "        # Return statistics\n",
    "        return float(loss), len(valloader.dataset), {\"accuracy\": float(accuracy)}\n",
    "\n",
    "\n",
    "def fit_config(server_round: int) -> Dict[str, Scalar]:\n",
    "    \"\"\"Return a configuration with static batch size and (local) epochs.\"\"\"\n",
    "    config = {\n",
    "        \"epochs\": 5,  # number of local epochs\n",
    "        \"batch_size\": 64,\n",
    "    }\n",
    "    return config\n",
    "\n",
    "\n",
    "def get_params(model: torch.nn.ModuleList) -> List[np.ndarray]:\n",
    "    \"\"\"Get model weights as a list of NumPy ndarrays.\"\"\"\n",
    "    return [val.cpu().numpy() for _, val in model.state_dict().items()]\n",
    "\n",
    "\n",
    "def set_params(model: torch.nn.ModuleList, params: List[np.ndarray]):\n",
    "    \"\"\"Set model weights from a list of NumPy ndarrays.\"\"\"\n",
    "    params_dict = zip(model.state_dict().keys(), params)\n",
    "    state_dict = OrderedDict({k: torch.from_numpy(np.copy(v)) for k, v in params_dict})\n",
    "    model.load_state_dict(state_dict, strict=True)\n",
    "\n",
    "\n",
    "def get_evaluate_fn(\n",
    "    testset: torchvision.datasets.FashionMNIST,\n",
    ") -> Callable[[fl.common.NDArrays], Optional[Tuple[float, float]]]:\n",
    "    \"\"\"Return an evaluation function for centralized evaluation.\"\"\"\n",
    "\n",
    "    def evaluate(\n",
    "        server_round: int, parameters: fl.common.NDArrays, config: Dict[str, Scalar]\n",
    "    ) -> Optional[Tuple[float, float]]:\n",
    "        \"\"\"Use the entire CIFAR-10 test set for evaluation.\"\"\"\n",
    "\n",
    "        # determine device\n",
    "        device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "        model = LeNet5()\n",
    "        set_params(model, parameters)\n",
    "        model.to(device)\n",
    "        \n",
    "        testloader = torch.utils.data.DataLoader(testset, batch_size=64, shuffle=True)\n",
    "        loss, accuracy = test(model, testloader, device=device)\n",
    "\n",
    "        # return statistics\n",
    "        return loss, {\"accuracy\": accuracy}\n",
    "\n",
    "    return evaluate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TorchVision_FL(VisionDataset):\n",
    "    \"\"\"This is just a trimmed down version of torchvision.datasets.MNIST.\n",
    "    Use this class by either passing a path to a torch file (.pt)\n",
    "    containing (data, targets) or pass the data, targets directly\n",
    "    instead.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        path_to_data=None,\n",
    "        data=None,\n",
    "        targets=None,\n",
    "        transform: Optional[Callable] = None,\n",
    "    ) -> None:\n",
    "        path = path_to_data.parent if path_to_data else None\n",
    "        super(TorchVision_FL, self).__init__(path, transform=transform)\n",
    "        self.transform = transform\n",
    "\n",
    "        if path_to_data:\n",
    "            # load data and targets (path_to_data points to an specific .pt file)\n",
    "            self.data, self.targets = torch.load(path_to_data)\n",
    "        else:\n",
    "            self.data = data\n",
    "            self.targets = targets\n",
    "\n",
    "    def __getitem__(self, index: int) -> Tuple[Any, Any]:\n",
    "        img, target = self.data[index], int(self.targets[index])\n",
    "\n",
    "        # doing this so that it is consistent with all other datasets\n",
    "        # to return a PIL Image\n",
    "        if not isinstance(img, Image.Image):  # if not PIL image\n",
    "            if not isinstance(img, np.ndarray):  # if torch tensor\n",
    "                img = img.numpy()\n",
    "\n",
    "            img = Image.fromarray(img)\n",
    "\n",
    "        if self.transform is not None:\n",
    "            img = self.transform(img)\n",
    "\n",
    "        if self.target_transform is not None:\n",
    "            target = self.target_transform(target)\n",
    "\n",
    "        return img, target\n",
    "\n",
    "    def __len__(self) -> int:\n",
    "        return len(self.data)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "pool_size = 100  # number of dataset partions (= number of total clients)\n",
    "\n",
    "client_resources = {\n",
    "        \"num_cpus\": cpu_count\n",
    "}  # each client will get allocated 1 CPUs\n",
    "\n",
    "transformations = transforms.Compose([\n",
    "    transforms.Resize((32,32)),\n",
    "    transforms.ToTensor(),\n",
    "])\n",
    " \n",
    "\n",
    "# Download Dataset\n",
    "try:\n",
    "    train_data = Food101(data_path, transform=transformations)\n",
    "except:\n",
    "    train_data = Food101(data_path, transform=transformations, download=True) \n",
    "test_data = Food101(data_path, split='test', transform=transformations)\n",
    "\n",
    "lengths = []\n",
    "while sum(lengths) != len(train_data):\n",
    "    lengths = [round(x) for x in np.random.dirichlet(\n",
    "        np.ones(pool_size),size=1)[0] * len(train_data)]\n",
    "\n",
    "client_datasets = torch.utils.data.random_split(train_data, lengths)\n",
    "\n",
    "for i, x in enumerate(client_datasets):\n",
    "    torch.save(x, os.path.join(data_path, 'client_%s' % str(i+1))) #human-readable clients starting w/client_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "101"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(test_data.classes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO flwr 2023-02-19 18:22:24,536 | app.py:145 | Starting Flower simulation, config: ServerConfig(num_rounds=100, round_timeout=None)\n",
      "INFO flwr 2023-02-19 18:22:26,646 | app.py:179 | Flower VCE: Ray initialized with resources: {'memory': 5845494990.0, 'node:192.168.1.171': 1.0, 'object_store_memory': 2922747494.0, 'CPU': 8.0}\n",
      "INFO flwr 2023-02-19 18:22:26,648 | server.py:86 | Initializing global parameters\n",
      "INFO flwr 2023-02-19 18:22:26,649 | server.py:270 | Requesting initial parameters from one random client\n",
      "2023-02-19 18:22:27,359\tERROR serialization.py:342 -- \n",
      "Traceback (most recent call last):\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 201, in _deserialize_msgpack_data\n",
      "    obj = MessagePackSerializer.loads(msgpack_data, _python_deserializer)\n",
      "  File \"python/ray/includes/serialization.pxi\", line 191, in ray._raylet.MessagePackSerializer.loads\n",
      "  File \"python/ray/includes/serialization.pxi\", line 192, in ray._raylet.MessagePackSerializer.loads\n",
      "  File \"msgpack/_unpacker.pyx\", line 161, in msgpack._unpacker.unpackb\n",
      "TypeError: unpackb() got an unexpected keyword argument 'strict_map_key'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 340, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 259, in _deserialize_object\n",
      "    obj = self._deserialize_msgpack_data(data, metadata_fields)\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 203, in _deserialize_msgpack_data\n",
      "    raise DeserializationError()\n",
      "ray.serialization.DeserializationError\n",
      "DEBUG flwr 2023-02-19 18:22:27,361 | ray_client_proxy.py:72 | System error: \n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 201, in _deserialize_msgpack_data\n",
      "    obj = MessagePackSerializer.loads(msgpack_data, _python_deserializer)\n",
      "  File \"python/ray/includes/serialization.pxi\", line 191, in ray._raylet.MessagePackSerializer.loads\n",
      "  File \"python/ray/includes/serialization.pxi\", line 192, in ray._raylet.MessagePackSerializer.loads\n",
      "  File \"msgpack/_unpacker.pyx\", line 161, in msgpack._unpacker.unpackb\n",
      "TypeError: unpackb() got an unexpected keyword argument 'strict_map_key'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 340, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 259, in _deserialize_object\n",
      "    obj = self._deserialize_msgpack_data(data, metadata_fields)\n",
      "  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 203, in _deserialize_msgpack_data\n",
      "    raise DeserializationError()\n",
      "ray.serialization.DeserializationError\n",
      "\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m 2023-02-19 18:22:27,356\tERROR serialization.py:342 -- \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 201, in _deserialize_msgpack_data\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     obj = MessagePackSerializer.loads(msgpack_data, _python_deserializer)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"python/ray/includes/serialization.pxi\", line 191, in ray._raylet.MessagePackSerializer.loads\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"python/ray/includes/serialization.pxi\", line 192, in ray._raylet.MessagePackSerializer.loads\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"msgpack/_unpacker.pyx\", line 161, in msgpack._unpacker.unpackb\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m TypeError: unpackb() got an unexpected keyword argument 'strict_map_key'\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m During handling of the above exception, another exception occurred:\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 340, in deserialize_objects\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     obj = self._deserialize_object(data, metadata, object_ref)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 237, in _deserialize_object\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 203, in _deserialize_msgpack_data\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     raise DeserializationError()\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m ray.serialization.DeserializationError\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m 2023-02-19 18:22:27,357\tERROR serialization.py:342 -- \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 201, in _deserialize_msgpack_data\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     obj = MessagePackSerializer.loads(msgpack_data, _python_deserializer)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"python/ray/includes/serialization.pxi\", line 191, in ray._raylet.MessagePackSerializer.loads\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"python/ray/includes/serialization.pxi\", line 192, in ray._raylet.MessagePackSerializer.loads\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"msgpack/_unpacker.pyx\", line 161, in msgpack._unpacker.unpackb\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m TypeError: unpackb() got an unexpected keyword argument 'strict_map_key'\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m During handling of the above exception, another exception occurred:\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 340, in deserialize_objects\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     obj = self._deserialize_object(data, metadata, object_ref)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 237, in _deserialize_object\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 203, in _deserialize_msgpack_data\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     raise DeserializationError()\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m ray.serialization.DeserializationError\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m 2023-02-19 18:22:27,357\tERROR serialization.py:342 -- \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 201, in _deserialize_msgpack_data\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     obj = MessagePackSerializer.loads(msgpack_data, _python_deserializer)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"python/ray/includes/serialization.pxi\", line 191, in ray._raylet.MessagePackSerializer.loads\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"python/ray/includes/serialization.pxi\", line 192, in ray._raylet.MessagePackSerializer.loads\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"msgpack/_unpacker.pyx\", line 161, in msgpack._unpacker.unpackb\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m TypeError: unpackb() got an unexpected keyword argument 'strict_map_key'\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m During handling of the above exception, another exception occurred:\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m \n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 340, in deserialize_objects\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     obj = self._deserialize_object(data, metadata, object_ref)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 237, in _deserialize_object\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m   File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 203, in _deserialize_msgpack_data\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m     raise DeserializationError()\n",
      "\u001b[2m\u001b[36m(launch_and_get_parameters pid=200550)\u001b[0m ray.serialization.DeserializationError\n"
     ]
    },
    {
     "ename": "RaySystemError",
     "evalue": "System error: \ntraceback: Traceback (most recent call last):\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 201, in _deserialize_msgpack_data\n    obj = MessagePackSerializer.loads(msgpack_data, _python_deserializer)\n  File \"python/ray/includes/serialization.pxi\", line 191, in ray._raylet.MessagePackSerializer.loads\n  File \"python/ray/includes/serialization.pxi\", line 192, in ray._raylet.MessagePackSerializer.loads\n  File \"msgpack/_unpacker.pyx\", line 161, in msgpack._unpacker.unpackb\nTypeError: unpackb() got an unexpected keyword argument 'strict_map_key'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 340, in deserialize_objects\n    obj = self._deserialize_object(data, metadata, object_ref)\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 259, in _deserialize_object\n    obj = self._deserialize_msgpack_data(data, metadata_fields)\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 203, in _deserialize_msgpack_data\n    raise DeserializationError()\nray.serialization.DeserializationError\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRaySystemError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-22fc8e72552f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;31m# start simulation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m fl.simulation.start_simulation(\n\u001b[0m\u001b[1;32m     20\u001b[0m     \u001b[0mclient_fn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mclient_fn\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m     \u001b[0mnum_clients\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpool_size\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/flwr/simulation/app.py\u001b[0m in \u001b[0;36mstart_simulation\u001b[0;34m(client_fn, num_clients, clients_ids, client_resources, server, config, strategy, client_manager, ray_init_args, keep_initialised)\u001b[0m\n\u001b[1;32m    194\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m     \u001b[0;31m# Start training\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m     hist = _fl(\n\u001b[0m\u001b[1;32m    197\u001b[0m         \u001b[0mserver\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minitialized_server\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m         \u001b[0mconfig\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minitialized_config\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/flwr/server/app.py\u001b[0m in \u001b[0;36m_fl\u001b[0;34m(server, config)\u001b[0m\n\u001b[1;32m    199\u001b[0m ) -> History:\n\u001b[1;32m    200\u001b[0m     \u001b[0;31m# Fit model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 201\u001b[0;31m     \u001b[0mhist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mserver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum_rounds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnum_rounds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mround_timeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    202\u001b[0m     \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mINFO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"app_fit: losses_distributed %s\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlosses_distributed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m     \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mINFO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"app_fit: metrics_distributed %s\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetrics_distributed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/flwr/server/server.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, num_rounds, timeout)\u001b[0m\n\u001b[1;32m     85\u001b[0m         \u001b[0;31m# Initialize parameters\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m         \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mINFO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Initializing global parameters\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_initial_parameters\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m         \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mINFO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Evaluating initial parameters\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrategy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mevaluate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/flwr/server/server.py\u001b[0m in \u001b[0;36m_get_initial_parameters\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    271\u001b[0m         \u001b[0mrandom_client\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_client_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    272\u001b[0m         \u001b[0mins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGetParametersIns\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 273\u001b[0;31m         \u001b[0mget_parameters_res\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrandom_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_parameters\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    274\u001b[0m         \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mINFO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Received initial parameters from one random client\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    275\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mget_parameters_res\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/flwr/simulation/ray_transport/ray_client_proxy.py\u001b[0m in \u001b[0;36mget_parameters\u001b[0;34m(self, ins, timeout)\u001b[0m\n\u001b[1;32m     71\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDEBUG\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 73\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     74\u001b[0m         return cast(\n\u001b[1;32m     75\u001b[0m             \u001b[0mcommon\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGetParametersRes\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/flwr/simulation/ray_transport/ray_client_proxy.py\u001b[0m in \u001b[0;36mget_parameters\u001b[0;34m(self, ins, timeout)\u001b[0m\n\u001b[1;32m     68\u001b[0m         ).remote(self.client_fn, self.cid, ins)\n\u001b[1;32m     69\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 70\u001b[0;31m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfuture_paramseters_res\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     71\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDEBUG\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/ray/_private/client_mode_hook.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    103\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"init\"\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mis_client_mode_enabled_by_default\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 105\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    107\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/py3data/lib/python3.8/site-packages/ray/worker.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(object_refs, timeout)\u001b[0m\n\u001b[1;32m   1831\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_instanceof_cause\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1832\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1833\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1834\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1835\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_individual_id\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRaySystemError\u001b[0m: System error: \ntraceback: Traceback (most recent call last):\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 201, in _deserialize_msgpack_data\n    obj = MessagePackSerializer.loads(msgpack_data, _python_deserializer)\n  File \"python/ray/includes/serialization.pxi\", line 191, in ray._raylet.MessagePackSerializer.loads\n  File \"python/ray/includes/serialization.pxi\", line 192, in ray._raylet.MessagePackSerializer.loads\n  File \"msgpack/_unpacker.pyx\", line 161, in msgpack._unpacker.unpackb\nTypeError: unpackb() got an unexpected keyword argument 'strict_map_key'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 340, in deserialize_objects\n    obj = self._deserialize_object(data, metadata, object_ref)\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 259, in _deserialize_object\n    obj = self._deserialize_msgpack_data(data, metadata_fields)\n  File \"/home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages/ray/serialization.py\", line 203, in _deserialize_msgpack_data\n    raise DeserializationError()\nray.serialization.DeserializationError\n"
     ]
    }
   ],
   "source": [
    "# configure the strategy\n",
    "strategy = fl.server.strategy.FedAvg(\n",
    "    fraction_fit=0.1,\n",
    "    fraction_evaluate=0.1,\n",
    "    min_fit_clients=10,\n",
    "    min_evaluate_clients=10,\n",
    "    min_available_clients=pool_size,  # All clients should be available\n",
    "    on_fit_config_fn=fit_config,\n",
    "    evaluate_fn=get_evaluate_fn(test_data),  # centralised evaluation of global model\n",
    ")\n",
    "\n",
    "def client_fn(cid: str):\n",
    "    # create a single client instance\n",
    "    return FlowerClient(cid, os.path.join(data_path, 'client_%s' % cid))\n",
    "\n",
    "ray_init_args = {\"include_dashboard\": False}\n",
    "\n",
    "# start simulation\n",
    "fl.simulation.start_simulation(\n",
    "    client_fn=client_fn,\n",
    "    num_clients=pool_size,\n",
    "    client_resources=client_resources,\n",
    "    config=fl.server.ServerConfig(num_rounds=100),\n",
    "    strategy=strategy\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import flwr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.3.0'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flwr.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting flwr\n",
      "  Downloading flwr-1.3.0-py3-none-any.whl (139 kB)\n",
      "\u001b[K     |████████████████████████████████| 139 kB 741 kB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied, skipping upgrade: iterators<0.0.3,>=0.0.2 in /home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages (from flwr) (0.0.2)\n",
      "Requirement already satisfied, skipping upgrade: numpy<2.0.0,>=1.21.0 in /home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages (from flwr) (1.23.4)\n",
      "Requirement already satisfied, skipping upgrade: grpcio!=1.52.0,<2.0.0,>=1.43.0 in /home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages (from flwr) (1.43.0)\n",
      "Requirement already satisfied, skipping upgrade: protobuf<4.0.0,>=3.19.0 in /home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages (from flwr) (3.20.3)\n",
      "Requirement already satisfied, skipping upgrade: six>=1.5.2 in /home/katharine/.virtualenvs/py3data/lib/python3.8/site-packages (from grpcio!=1.52.0,<2.0.0,>=1.43.0->flwr) (1.16.0)\n",
      "Installing collected packages: flwr\n",
      "  Attempting uninstall: flwr\n",
      "    Found existing installation: flwr 1.0.0\n",
      "    Uninstalling flwr-1.0.0:\n",
      "      Successfully uninstalled flwr-1.0.0\n",
      "Successfully installed flwr-1.3.0\n"
     ]
    }
   ],
   "source": [
    "!pip install --upgrade flwr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Challenges\n",
    "\n",
    "- Adjust the fit and evaluate settings and see how the performance changes.\n",
    "- Try out another [Flower tutorial](https://flower.dev/docs/quickstart-pytorch.html).\n",
    "- Get a group of several folks together to try running flower in a distributed setup. Document your learnings and share in the reader-contributions!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
